File: packages/js-dpp/lib/stateTransition/StateTransitionFacade.js
const $RefParser = require('@apidevtools/json-schema-ref-parser');
const jsonPatch = require('fast-json-patch');
const jsonSchemaDiffValidator = require('json-schema-diff-validator');
const { Signer: { verifyHashSignature } } = require('@dashevo/dashcore-lib');

const MissingOptionError = require('../errors/MissingOptionError');

const StateTransitionFactory = require('./StateTransitionFactory');

const AbstractStateTransition = require('./AbstractStateTransition');

const stateTransitionTypes = require('./stateTransitionTypes');

const createStateTransitionFactory = require('./createStateTransitionFactory');

const validateDataContractFactory = require('../dataContract/validation/validateDataContractFactory');
const validateDataContractPatternsFactory = require('../dataContract/validation/validateDataContractPatternsFactory');
const validateDataContractCreateTransitionBasicFactory = require('../dataContract/stateTransition/DataContractCreateTransition/validation/basic/validateDataContractCreateTransitionBasicFactory');
const validateStateTransitionBasicFactory = require('./validation/validateStateTransitionBasicFactory');
const validateDataContractCreateTransitionStateFactory = require('../dataContract/stateTransition/DataContractCreateTransition/validation/state/validateDataContractCreateTransitionStateFactory');
const validateStateTransitionStateFactory = require('./validation/validateStateTransitionStateFactory');
const validateDocumentsBatchTransitionBasicFactory = require('../document/stateTransition/DocumentsBatchTransition/validation/basic/validateDocumentsBatchTransitionBasicFactory');
const validateIdentityCreateTransitionStateFactory = require('../identity/stateTransition/IdentityCreateTransition/validation/state/validateIdentityCreateTransitionStateFactory');
const validateIdentityTopUpTransitionStateFactory = require('../identity/stateTransition/IdentityTopUpTransition/validation/state/validateIdentityTopUpTransitionStateFactory');
const validateIdentityUpdateTransitionStateFactory = require('../identity/stateTransition/IdentityUpdateTransition/validation/state/validateIdentityUpdateTransitionStateFactory');
const validateIdentityCreateTransitionBasicFactory = require('../identity/stateTransition/IdentityCreateTransition/validation/basic/validateIdentityCreateTransitionBasicFactory');
const validateIdentityTopUpTransitionBasicFactory = require('../identity/stateTransition/IdentityTopUpTransition/validation/basic/validateIdentityTopUpTransitionBasicFactory');
const validateIdentityUpdateTransitionBasicFactory = require('../identity/stateTransition/IdentityUpdateTransition/validation/basic/validateIdentityUpdateTransitionBasicFactory');
const validateStateTransitionIdentitySignatureFactory = require('./validation/validateStateTransitionIdentitySignatureFactory');
const validateStateTransitionFeeFactory = require('./validation/validateStateTransitionFeeFactory');

const enrichDataContractWithBaseSchema = require('../dataContract/enrichDataContractWithBaseSchema');
const findDuplicatesById = require('../document/stateTransition/DocumentsBatchTransition/validation/basic/findDuplicatesById');
const findDuplicatesByIndices = require('../document/stateTransition/DocumentsBatchTransition/validation/basic/findDuplicatesByIndices');

const validateDocumentsBatchTransitionStateFactory = require('../document/stateTransition/DocumentsBatchTransition/validation/state/validateDocumentsBatchTransitionStateFactory');
const fetchDocumentsFactory = require('../document/stateTransition/DocumentsBatchTransition/validation/state/fetchDocumentsFactory');
const validateDocumentsUniquenessByIndicesFactory = require('../document/stateTransition/DocumentsBatchTransition/validation/state/validateDocumentsUniquenessByIndicesFactory');
const validatePartialCompoundIndices = require('../document/stateTransition/DocumentsBatchTransition/validation/basic/validatePartialCompoundIndices');
const getDataTriggersFactory = require('../dataTrigger/getDataTriggersFactory');
const executeDataTriggersFactory = require('../document/stateTransition/DocumentsBatchTransition/validation/state/executeDataTriggersFactory');
const validatePublicKeysFactory = require('../identity/validation/validatePublicKeysFactory');
const validatePublicKeysState = require('../identity/stateTransition/IdentityUpdateTransition/validation/state/validatePublicKeysState');
const validateRequiredPurposeAndSecurityLevelFactory = require('../identity/validation/validateRequiredPurposeAndSecurityLevelFactory');
const validateDataContractMaxDepthFactory = require('../dataContract/validation/validateDataContractMaxDepthFactory');

const applyStateTransitionFactory = require('./applyStateTransitionFactory');

const applyDataContractCreateTransitionFactory = require(
  '../dataContract/stateTransition/DataContractCreateTransition/applyDataContractCreateTransitionFactory',
);
const applyDataContractUpdateTransitionFactory = require(
  '../dataContract/stateTransition/DataContractUpdateTransition/applyDataContractUpdateTransitionFactory',
);

const applyDocumentsBatchTransitionFactory = require(
  '../document/stateTransition/DocumentsBatchTransition/applyDocumentsBatchTransitionFactory',
);

const applyIdentityCreateTransitionFactory = require(
  '../identity/stateTransition/IdentityCreateTransition/applyIdentityCreateTransitionFactory',
);

const applyIdentityTopUpTransitionFactory = require(
  '../identity/stateTransition/IdentityTopUpTransition/applyIdentityTopUpTransitionFactory',
);

const applyIdentityUpdateTransitionFactory = require(
  '../identity/stateTransition/IdentityUpdateTransition/applyIdentityUpdateTransitionFactory',
);
const validateInstantAssetLockProofStructureFactory = require('../identity/stateTransition/assetLockProof/instant/validateInstantAssetLockProofStructureFactory');
const calculateStateTransitionFee = require('./fee/calculateStateTransitionFee');
const InstantAssetLockProof = require('../identity/stateTransition/assetLockProof/instant/InstantAssetLockProof');
const ChainAssetLockProof = require('../identity/stateTransition/assetLockProof/chain/ChainAssetLockProof');
const validateChainAssetLockProofStructureFactory = require('../identity/stateTransition/assetLockProof/chain/validateChainAssetLockProofStructureFactory');
const fetchAssetLockTransactionOutputFactory = require('../identity/stateTransition/assetLockProof/fetchAssetLockTransactionOutputFactory');
const validateAssetLockTransactionFactory = require('../identity/stateTransition/assetLockProof/validateAssetLockTransactionFactory');

const ValidationResult = require('../validation/ValidationResult');
const AbstractStateTransitionIdentitySigned = require('./AbstractStateTransitionIdentitySigned');
const validateStateTransitionKeySignatureFactory = require('./validation/validateStateTransitionKeySignatureFactory');

const fetchAssetLockPublicKeyHashFactory = require('../identity/stateTransition/assetLockProof/fetchAssetLockPublicKeyHashFactory');

const decodeProtocolEntityFactory = require('../decodeProtocolEntityFactory');
const protocolVersion = require('../version/protocolVersion');
const validateProtocolVersionFactory = require('../version/validateProtocolVersionFactory');
const validateDataContractUpdateTransitionBasicFactory = require('../dataContract/stateTransition/DataContractUpdateTransition/validation/basic/validateDataContractUpdateTransitionBasicFactory');
const validateDataContractUpdateTransitionStateFactory = require('../dataContract/stateTransition/DataContractUpdateTransition/validation/state/validateDataContractUpdateTransitionStateFactory');
const validateIndicesAreBackwardCompatible = require('../dataContract/stateTransition/DataContractUpdateTransition/validation/basic/validateIndicesAreBackwardCompatible');
const getPropertyDefinitionByPath = require('../dataContract/getPropertyDefinitionByPath');

const identityJsonSchema = require('../../schema/identity/stateTransition/publicKey.json');
const validatePublicKeySignaturesFactory = require('../identity/stateTransition/validatePublicKeySignaturesFactory');
const StateTransitionExecutionContext = require('./StateTransitionExecutionContext');

class StateTransitionFacade {
  /**
   * @param {DashPlatformProtocol} dpp
   * @param {RE2} RE2
   * @param {BlsSignatures} bls
   */
  constructor(dpp, RE2, bls) {
    this.stateRepository = dpp.getStateRepository();

    const validator = dpp.getJsonSchemaValidator();

    const validateDataContractMaxDepth = validateDataContractMaxDepthFactory($RefParser);
    const validateDataContractPatterns = validateDataContractPatternsFactory(RE2);

    const validateProtocolVersion = validateProtocolVersionFactory(
      dpp,
      protocolVersion.compatibility,
    );
    const validateDataContract = validateDataContractFactory(
      validator,
      validateDataContractMaxDepth,
      enrichDataContractWithBaseSchema,
      validateDataContractPatterns,
      validateProtocolVersion,
      getPropertyDefinitionByPath,
    );

    this.validateStateTransitionIdentitySignature = validateStateTransitionIdentitySignatureFactory(
      this.stateRepository,
    );

    const fetchAssetLockTransactionOutput = fetchAssetLockTransactionOutputFactory(
      this.stateRepository,
    );

    const fetchAssetLockPublicKeyHash = fetchAssetLockPublicKeyHashFactory(
      fetchAssetLockTransactionOutput,
    );

    this.validateStateTransitionKeySignature = validateStateTransitionKeySignatureFactory(
      verifyHashSignature,
      fetchAssetLockPublicKeyHash,
    );

    // eslint-disable-next-line max-len
    const validateDataContractCreateTransitionBasic = validateDataContractCreateTransitionBasicFactory(
      validator,
      validateDataContract,
      validateProtocolVersion,
    );

    // eslint-disable-next-line max-len
    const validateDataContractUpdateTransitionBasic = validateDataContractUpdateTransitionBasicFactory(
      validator,
      validateDataContract,
      validateProtocolVersion,
      this.stateRepository,
      jsonSchemaDiffValidator,
      validateIndicesAreBackwardCompatible,
      jsonPatch,
    );

    this.createStateTransition = createStateTransitionFactory(this.stateRepository);

    const validateDocumentsBatchTransitionBasic = (
      validateDocumentsBatchTransitionBasicFactory(
        findDuplicatesById,
        findDuplicatesByIndices,
        this.stateRepository,
        validator,
        enrichDataContractWithBaseSchema,
        validatePartialCompoundIndices,
        validateProtocolVersion,
      )
    );

    const validateAssetLockTransaction = validateAssetLockTransactionFactory(this.stateRepository);

    const validateInstantAssetLockProofStructure = validateInstantAssetLockProofStructureFactory(
      validator,
      this.stateRepository,
      validateAssetLockTransaction,
    );

    const validateChainAssetLockProofStructure = validateChainAssetLockProofStructureFactory(
      validator,
      this.stateRepository,
      validateAssetLockTransaction,
    );

    const proofValidationFunctionsByType = {
      [InstantAssetLockProof.type]: validateInstantAssetLockProofStructure,
      [ChainAssetLockProof.type]: validateChainAssetLockProofStructure,
    };

    const validatePublicKeys = validatePublicKeysFactory(
      validator,
      identityJsonSchema,
      bls,
    );

    const validateRequiredPurposeAndSecurityLevel = (
      validateRequiredPurposeAndSecurityLevelFactory()
    );

    const validatePublicKeySignatures = validatePublicKeySignaturesFactory(
      this.createStateTransition,
    );

    const validateIdentityCreateTransitionBasic = (
      validateIdentityCreateTransitionBasicFactory(
        validator,
        validatePublicKeys,
        validateRequiredPurposeAndSecurityLevel,
        proofValidationFunctionsByType,
        validateProtocolVersion,
        validatePublicKeySignatures,
      )
    );

    const validateIdentityTopUpTransitionBasic = (
      validateIdentityTopUpTransitionBasicFactory(
        validator,
        proofValidationFunctionsByType,
        validateProtocolVersion,
      )
    );

    const validateIdentityUpdateTransitionBasic = validateIdentityUpdateTransitionBasicFactory(
      validator,
      validateProtocolVersion,
      validatePublicKeys,
      validatePublicKeySignatures,
    );

    const validationFunctionsByType = {
      [stateTransitionTypes.DATA_CONTRACT_CREATE]: validateDataContractCreateTransitionBasic,
      [stateTransitionTypes.DATA_CONTRACT_UPDATE]: validateDataContractUpdateTransitionBasic,
      [stateTransitionTypes.DOCUMENTS_BATCH]: validateDocumentsBatchTransitionBasic,
      [stateTransitionTypes.IDENTITY_CREATE]: validateIdentityCreateTransitionBasic,
      [stateTransitionTypes.IDENTITY_TOP_UP]: validateIdentityTopUpTransitionBasic,
      [stateTransitionTypes.IDENTITY_UPDATE]: validateIdentityUpdateTransitionBasic,
    };

    this.validateStateTransitionBasic = validateStateTransitionBasicFactory(
      validationFunctionsByType,
      this.createStateTransition,
    );

    const validateDataContractCreateTransitionState = (
      validateDataContractCreateTransitionStateFactory(
        this.stateRepository,
      )
    );

    const validateDataContractUpdateTransitionState = (
      validateDataContractUpdateTransitionStateFactory(
        this.stateRepository,
      )
    );

    const validateIdentityCreateTransitionState = validateIdentityCreateTransitionStateFactory(
      this.stateRepository,
    );

    const validateIdentityTopUpTransitionState = validateIdentityTopUpTransitionStateFactory();

    const validateIdentityUpdateTransitionState = validateIdentityUpdateTransitionStateFactory(
      this.stateRepository,
      validatePublicKeysState,
      validateRequiredPurposeAndSecurityLevel,
    );

    const fetchDocuments = fetchDocumentsFactory(
      this.stateRepository,
    );

    const validateDocumentsUniquenessByIndices = validateDocumentsUniquenessByIndicesFactory(
      this.stateRepository,
    );

    const getDataTriggers = getDataTriggersFactory();

    const executeDataTriggers = executeDataTriggersFactory(
      getDataTriggers,
    );

    const validateDocumentsBatchTransitionState = validateDocumentsBatchTransitionStateFactory(
      this.stateRepository,
      fetchDocuments,
      validateDocumentsUniquenessByIndices,
      executeDataTriggers,
    );

    this.validateStateTransitionState = validateStateTransitionStateFactory({
      [stateTransitionTypes.DATA_CONTRACT_CREATE]: validateDataContractCreateTransitionState,
      [stateTransitionTypes.DATA_CONTRACT_UPDATE]: validateDataContractUpdateTransitionState,
      [stateTransitionTypes.DOCUMENTS_BATCH]: validateDocumentsBatchTransitionState,
      [stateTransitionTypes.IDENTITY_CREATE]: validateIdentityCreateTransitionState,
      [stateTransitionTypes.IDENTITY_TOP_UP]: validateIdentityTopUpTransitionState,
      [stateTransitionTypes.IDENTITY_UPDATE]: validateIdentityUpdateTransitionState,
    });

    this.validateStateTransitionFee = validateStateTransitionFeeFactory(
      this.stateRepository,
      calculateStateTransitionFee,
      fetchAssetLockTransactionOutput,
    );

    const decodeProtocolEntity = decodeProtocolEntityFactory();

    this.factory = new StateTransitionFactory(
      this.validateStateTransitionBasic,
      this.createStateTransition,
      dpp,
      decodeProtocolEntity,
    );

    const applyDataContractCreateTransition = applyDataContractCreateTransitionFactory(
      this.stateRepository,
    );

    const applyDataContractUpdateTransition = applyDataContractUpdateTransitionFactory(
      this.stateRepository,
    );

    const applyDocumentsBatchTransition = applyDocumentsBatchTransitionFactory(
      this.stateRepository,
      fetchDocuments,
    );

    const applyIdentityCreateTransition = applyIdentityCreateTransitionFactory(
      this.stateRepository,
      fetchAssetLockTransactionOutput,
    );

    const applyIdentityTopUpTransition = applyIdentityTopUpTransitionFactory(
      this.stateRepository,
      fetchAssetLockTransactionOutput,
    );

    const applyIdentityUpdateTransition = applyIdentityUpdateTransitionFactory(
      this.stateRepository,
    );

    this.applyStateTransition = applyStateTransitionFactory(
      applyDataContractCreateTransition,
      applyDataContractUpdateTransition,
      applyDocumentsBatchTransition,
      applyIdentityCreateTransition,
      applyIdentityTopUpTransition,
      applyIdentityUpdateTransition,
    );
  }

  /**
   * Create State Transition from plain object
   *
   * @param {RawDataContractCreateTransition|RawDocumentsBatchTransition} rawStateTransition
   * @param {Object} [options]
   * @param {boolean} [options.skipValidation=false]
   * @return {DataContractCreateTransition|DocumentsBatchTransition}
   */
  async createFromObject(rawStateTransition, options = {}) {
    if (!this.stateRepository && !options.skipValidation) {
      throw new MissingOptionError(
        'stateRepository',
        'Can\'t create State Transition because State Repository is not set, use'
        + ' setStateRepository method',
      );
    }

    return this.factory.createFromObject(rawStateTransition, options);
  }

  /**
   * Create State Transition from buffer
   *
   * @param {Buffer} buffer
   * @param {Object} [options]
   * @param {boolean} [options.skipValidation=false]
   * @return {DataContractCreateTransition|DocumentsBatchTransition}
   */
  async createFromBuffer(buffer, options = {}) {
    if (!this.stateRepository && !options.skipValidation) {
      throw new MissingOptionError(
        'stateRepository',
        'Can\'t create State Transition because State Repository is not set, use'
        + ' setStateRepository method',
      );
    }

    return this.factory.createFromBuffer(buffer, options);
  }

  /**
   * Validate State Transition
   *
   * @param {RawStateTransition|AbstractStateTransition} stateTransition
   * @param {Object} [options]
   * @param {boolean} [options.basic=true]
   * @param {boolean} [options.signature=true]
   * @param {boolean} [options.fee=true]
   * @param {boolean} [options.state=true]
   * @return {Promise<ValidationResult>}
   */
  async validate(stateTransition, options = {}) {
    // eslint-disable-next-line no-param-reassign
    options = {
      basic: true,
      signature: true,
      fee: true,
      state: true,
      ...options,
    };

    if (!this.stateRepository) {
      throw new MissingOptionError(
        'stateRepository',
        'Can\'t validate State Transition because State Repository is not set, use'
        + ' setStateRepository method',
      );
    }

    // Convert raw state transition to the model
    let stateTransitionModel = stateTransition;

    if (!(stateTransition instanceof AbstractStateTransition)) {
      stateTransitionModel = await this.createStateTransition(stateTransition);
    }

    const result = new ValidationResult();

    // Basic validation
    if (options.basic) {
      result.merge(
        await this.validateBasic(stateTransition),
      );
    }

    if (!result.isValid()) {
      return result;
    }

    // Signature validation
    if (options.signature) {
      result.merge(
        await this.validateSignature(stateTransitionModel),
      );
    }

    if (!result.isValid()) {
      return result;
    }

    // Fee validation
    if (options.fee) {
      result.merge(
        await this.validateFee(stateTransitionModel),
      );
    }

    if (!result.isValid()) {
      return result;
    }

    // Validate against existing state
    if (options.state) {
      result.merge(
        await this.validateState(stateTransitionModel),
      );
    }

    return result;
  }

  /**
   * Validate State Transition structure and data
   *
   * @param {AbstractStateTransition|RawStateTransition} stateTransition
   * @return {Promise<ValidationResult>}
   */
  async validateBasic(stateTransition) {
    if (!this.stateRepository) {
      throw new MissingOptionError(
        'stateRepository',
        'Can\'t validate State Transition because State Repository is not set, use'
        + ' setStateRepository method',
      );
    }

    let rawStateTransition;
    let executionContext;

    if (stateTransition instanceof AbstractStateTransition) {
      rawStateTransition = stateTransition.toObject();
      executionContext = stateTransition.getExecutionContext();
    } else {
      rawStateTransition = stateTransition;
    }

    return this.validateStateTransitionBasic(
      rawStateTransition,
      executionContext || new StateTransitionExecutionContext(),
    );
  }

  /**
   * Validate State Transition signature and ownership
   *
   * @param {AbstractStateTransition} stateTransition
   *
   * @return {Promise<ValidationResult>}
   */
  async validateSignature(stateTransition) {
    if (!this.stateRepository) {
      throw new MissingOptionError(
        'stateRepository',
        'Can\'t validate State Transition because State Repository is not set, use'
        + ' setStateRepository method',
      );
    }

    if (stateTransition instanceof AbstractStateTransitionIdentitySigned) {
      return this.validateStateTransitionIdentitySignature(stateTransition);
    }

    return this.validateStateTransitionKeySignature(stateTransition);
  }

  /**
   * Validate State Transition fee
   *
   * @param {AbstractStateTransition} stateTransition
   *
   * @return {Promise<ValidationResult>}
   */
  async validateFee(stateTransition) {
    if (!this.stateRepository) {
      throw new MissingOptionError(
        'stateRepository',
        'Can\'t validate State Transition because State Repository is not set, use'
        + ' setStateRepository method',
      );
    }

    return this.validateStateTransitionFee(stateTransition);
  }

  /**
   * Validate State Transition against existing state
   *
   * @param {AbstractStateTransition} stateTransition
   *
   * @return {Promise<ValidationResult>}
   */
  async validateState(stateTransition) {
    if (!this.stateRepository) {
      throw new MissingOptionError(
        'stateRepository',
        'Can\'t validate State Transition because State Repository is not set, use'
        + ' setStateRepository method',
      );
    }

    return this.validateStateTransitionState(stateTransition);
  }

  /**
   * Apply state transition to the state
   *
   * @param {AbstractStateTransition} stateTransition
   *
   * @return {Promise<void>}
   */
  async apply(stateTransition) {
    return this.applyStateTransition(stateTransition);
  }
}

module.exports = StateTransitionFacade;

File: packages/rs-dpp/src/old/state_transition_facade.rs
use std::ops::Deref;

use std::sync::Arc;
use platform_value::Value;
use crate::{BlsModule, ProtocolError};
use crate::data_contract::state_transition::data_contract_create_transition::validation::state::validate_data_contract_create_transition_basic::DataContractCreateTransitionBasicValidator;
use crate::data_contract::state_transition::data_contract_update_transition::validation::basic::DataContractUpdateTransitionBasicValidator;
use crate::document::validation::basic::validate_documents_batch_transition_basic::DocumentBatchTransitionBasicValidator;
use crate::identity::state_transition::asset_lock_proof::{AssetLockProofValidator, AssetLockPublicKeyHashFetcher, AssetLockTransactionOutputFetcher, AssetLockTransactionValidator, ChainAssetLockProofStructureValidator, InstantAssetLockProofStructureValidator};
use crate::identity::state_transition::identity_create_transition::validation::basic::IdentityCreateTransitionBasicValidator;
use crate::identity::state_transition::identity_credit_withdrawal_transition::validation::basic::validate_identity_credit_withdrawal_transition_basic::IdentityCreditWithdrawalTransitionBasicValidator;
use crate::identity::state_transition::identity_topup_transition::validation::basic::IdentityTopUpTransitionBasicValidator;
use crate::identity::state_transition::identity_update_transition::validate_identity_update_transition_basic::ValidateIdentityUpdateTransitionBasic;
use crate::identity::state_transition::identity_credit_transfer_transition::validation::basic::identity_credit_transfer_basic::IdentityCreditTransferTransitionBasicValidator;
use crate::identity::state_transition::validate_public_key_signatures::PublicKeysSignaturesValidator;
use crate::identity::validation::{PUBLIC_KEY_SCHEMA_FOR_TRANSITION, PublicKeysValidator};

use crate::state_repository::StateRepositoryLike;
use crate::state_transition::apply_state_transition::ApplyStateTransition;
use crate::state_transition::{
    StateTransition, StateTransitionAction, StateTransitionFactory, StateTransitionFactoryOptions,
    StateTransitionFieldTypes,
};

use crate::state_transition::state_transition_execution_context::StateTransitionExecutionContext;
use crate::state_transition::validation::validate_state_transition_basic::StateTransitionBasicValidator;
use crate::state_transition::validation::validate_state_transition_by_type::StateTransitionByTypeValidator;
use crate::state_transition::validation::validate_state_transition_fee::StateTransitionFeeValidator;
use crate::state_transition::validation::validate_state_transition_identity_signature::validate_state_transition_identity_signature;
use crate::state_transition::validation::validate_state_transition_key_signature::StateTransitionKeySignatureValidator;
use crate::state_transition::validation::validate_state_transition_state::StateTransitionStateValidator;
use crate::validation::{
    AsyncDataValidator, AsyncDataValidatorWithContext, ConsensusValidationResult,
    SimpleConsensusValidationResult,
};
use crate::version::ProtocolVersionValidator;

#[derive(Clone)]
pub struct StateTransitionFacade<SR, BLS>
where
    SR: StateRepositoryLike + Clone,
    BLS: BlsModule + Clone,
{
    state_repository: Arc<SR>,
    basic_validator:
        Arc<StateTransitionBasicValidator<SR, StateTransitionByTypeValidator<SR, BLS>>>,
    state_validator: Arc<StateTransitionStateValidator<SR>>,
    key_signature_validator: Arc<StateTransitionKeySignatureValidator<SR>>,
    fee_validator: Arc<StateTransitionFeeValidator<SR>>,
    bls: BLS,
    factory: Arc<StateTransitionFactory<SR, BLS>>,
    apply_state_transition: ApplyStateTransition<SR>,
}

impl<SR, BLS> StateTransitionFacade<SR, BLS>
where
    SR: StateRepositoryLike + Clone,
    BLS: BlsModule + Clone,
{
    pub fn new(
        state_repository: SR,
        adapter: BLS,
        protocol_version_validator: Arc<ProtocolVersionValidator>,
    ) -> Result<Self, ProtocolError> {
        let wrapped_state_repository = Arc::new(state_repository.clone());

        let state_transition_basic_validator = {
            let pk_validator = Arc::new(
                PublicKeysValidator::new_with_schema(
                    PUBLIC_KEY_SCHEMA_FOR_TRANSITION.clone(),
                    adapter.clone(),
                )
                .map_err(|_| {
                    ProtocolError::Generic(String::from("Unable to initialize PublicKeysValidator"))
                })?,
            );
            let pk_sig_validator = Arc::new(PublicKeysSignaturesValidator::new(adapter.clone()));

            let asset_lock_tx_validator = Arc::new(AssetLockTransactionValidator::new(
                wrapped_state_repository.clone(),
            ));

            let asset_lock_validator = Arc::new(AssetLockProofValidator::new(
                InstantAssetLockProofStructureValidator::new(
                    wrapped_state_repository.clone(),
                    asset_lock_tx_validator.clone(),
                )
                .map_err(|_| {
                    ProtocolError::Generic(String::from(
                        "Unable to initialize InstantAssetLockProofStructureValidator",
                    ))
                })?,
                ChainAssetLockProofStructureValidator::new(
                    wrapped_state_repository.clone(),
                    asset_lock_tx_validator,
                )
                .map_err(|_| {
                    ProtocolError::Generic(String::from(
                        "Unable to initialize ChainAssetLockProofStructureValidator",
                    ))
                })?,
            ));

            let validator = StateTransitionBasicValidator::new(
                wrapped_state_repository.clone(),
                StateTransitionByTypeValidator::new(
                    DataContractCreateTransitionBasicValidator::new(
                        protocol_version_validator.clone(),
                    )?,
                    DataContractUpdateTransitionBasicValidator::new(
                        wrapped_state_repository.clone(),
                        protocol_version_validator.clone(),
                    )
                    .map_err(|_| {
                        ProtocolError::Generic(String::from(
                            "Unable to initialize DataContractUpdateTransitionBasicValidator",
                        ))
                    })?,
                    IdentityCreateTransitionBasicValidator::new(
                        protocol_version_validator.deref().clone(),
                        pk_validator.clone(),
                        pk_validator.clone(),
                        asset_lock_validator.clone(),
                        adapter.clone(),
                        pk_sig_validator.clone(),
                    )
                    .map_err(|_| {
                        ProtocolError::Generic(String::from(
                            "Unable to initialize IdentityCreateTransitionBasicValidator",
                        ))
                    })?,
                    ValidateIdentityUpdateTransitionBasic::new(
                        ProtocolVersionValidator::default(),
                        pk_validator,
                        pk_sig_validator,
                    )?,
                    IdentityTopUpTransitionBasicValidator::new(
                        ProtocolVersionValidator::default(),
                        asset_lock_validator,
                    )
                    .map_err(|_| {
                        ProtocolError::Generic(String::from(
                            "Unable to initialize IdentityTopUpTransitionBasicValidator",
                        ))
                    })?,
                    IdentityCreditWithdrawalTransitionBasicValidator::new(
                        protocol_version_validator.clone(),
                    )
                    .map_err(|_| {
                        ProtocolError::Generic(String::from(
                            "Unable to initialize IdentityCreditWithdrawalTransitionBasicValidator",
                        ))
                    })?,
                    DocumentBatchTransitionBasicValidator::new(
                        wrapped_state_repository.clone(),
                        protocol_version_validator.clone(),
                    ),
                    IdentityCreditTransferTransitionBasicValidator::new(
                        ProtocolVersionValidator::default(),
                    )
                    .map_err(|_| {
                        ProtocolError::Generic(String::from(
                            "Unable to initialize IdentityCreditTransferTransitionBasicValidator",
                        ))
                    })?,
                ),
            );

            Arc::new(validator)
        };

        let state_transition_factory = StateTransitionFactory::new(
            wrapped_state_repository.clone(),
            state_transition_basic_validator.clone(),
        );

        let asset_lock_transaction_output_fetcher = Arc::new(
            AssetLockTransactionOutputFetcher::new(wrapped_state_repository.clone()),
        );

        let state_transition_key_signature_validator = {
            let asset_public_key_hash_fetcher = AssetLockPublicKeyHashFetcher::new(
                wrapped_state_repository.clone(),
                asset_lock_transaction_output_fetcher.clone(),
            );

            StateTransitionKeySignatureValidator::new(
                wrapped_state_repository.clone(),
                asset_public_key_hash_fetcher,
            )
        };

        let state_transition_fee_validator =
            StateTransitionFeeValidator::new(wrapped_state_repository.clone());

        let state_transition_state_validator = StateTransitionStateValidator::new(state_repository);

        Ok(Self {
            state_repository: wrapped_state_repository.clone(),
            factory: Arc::new(state_transition_factory),
            basic_validator: state_transition_basic_validator,
            key_signature_validator: Arc::new(state_transition_key_signature_validator),
            fee_validator: Arc::new(state_transition_fee_validator),
            state_validator: Arc::new(state_transition_state_validator),
            bls: adapter,
            apply_state_transition: ApplyStateTransition::new(
                wrapped_state_repository,
                asset_lock_transaction_output_fetcher,
            ),
        })
    }

    pub async fn create_from_object(
        &self,
        state_transition: Value,
        skip_validation: bool,
    ) -> Result<StateTransition, ProtocolError> {
        self.factory
            .create_from_object(
                state_transition,
                Some(StateTransitionFactoryOptions { skip_validation }),
            )
            .await
    }

    pub async fn create_from_buffer(
        &self,
        buffer: &[u8],
        skip_validation: bool,
    ) -> Result<StateTransition, ProtocolError> {
        self.factory
            .create_from_buffer(
                buffer,
                Some(StateTransitionFactoryOptions { skip_validation }),
            )
            .await
    }

    pub async fn validate(
        &self,
        state_transition: &StateTransition,
        execution_context: &StateTransitionExecutionContext,
        options: ValidateOptions,
    ) -> Result<ConsensusValidationResult<Option<StateTransitionAction>>, ProtocolError> {
        let mut result =
            ConsensusValidationResult::<Option<StateTransitionAction>>::new_with_data(None);
        if options.basic {
            let state_transition_cleaned = state_transition.to_cleaned_object(false)?;
            result.merge(
                self.validate_basic(&state_transition_cleaned, execution_context)
                    .await?,
            );

            if !result.is_valid() {
                return Ok(result);
            }
        }

        if options.signature {
            result.merge(
                self.validate_signature(state_transition.clone(), execution_context)
                    .await?,
            );

            if !result.is_valid() {
                return Ok(result);
            }
        }

        if options.fee {
            result.merge(
                self.validate_fee(state_transition, execution_context)
                    .await?,
            );

            if !result.is_valid() {
                return Ok(result);
            }
        }

        if options.state {
            Ok(self
                .validate_state(state_transition, execution_context)
                .await?
                .map(Some))
        } else {
            Ok(result)
        }
    }

    pub async fn validate_basic(
        &self,
        state_transition: &Value,
        execution_context: &StateTransitionExecutionContext,
    ) -> Result<SimpleConsensusValidationResult, ProtocolError> {
        self.basic_validator
            .validate(state_transition, execution_context)
            .await
    }

    pub async fn validate_signature(
        &self,
        mut state_transition: StateTransition,
        execution_context: &StateTransitionExecutionContext,
    ) -> Result<SimpleConsensusValidationResult, ProtocolError> {
        // TODO: can we avoid duplicated code here?
        match state_transition {
            StateTransition::DataContractCreate(ref mut st) => {
                validate_state_transition_identity_signature(
                    self.state_repository.clone(),
                    st,
                    &self.bls,
                    execution_context,
                )
                .await
            }
            StateTransition::DataContractUpdate(ref mut st) => {
                validate_state_transition_identity_signature(
                    self.state_repository.clone(),
                    st,
                    &self.bls,
                    execution_context,
                )
                .await
            }
            StateTransition::DocumentsBatch(ref mut st) => {
                validate_state_transition_identity_signature(
                    self.state_repository.clone(),
                    st,
                    &self.bls,
                    execution_context,
                )
                .await
            }
            StateTransition::IdentityCreditWithdrawal(ref mut st) => {
                validate_state_transition_identity_signature(
                    self.state_repository.clone(),
                    st,
                    &self.bls,
                    execution_context,
                )
                .await
            }
            StateTransition::IdentityUpdate(ref mut st) => {
                validate_state_transition_identity_signature(
                    self.state_repository.clone(),
                    st,
                    &self.bls,
                    execution_context,
                )
                .await
            }
            _ => {
                self.key_signature_validator
                    .validate(&state_transition, execution_context)
                    .await
            }
        }
    }

    pub async fn validate_fee(
        &self,
        state_transition: &StateTransition,
        execution_context: &StateTransitionExecutionContext,
    ) -> Result<SimpleConsensusValidationResult, ProtocolError> {
        self.fee_validator
            .validate(state_transition, execution_context)
            .await
    }

    pub async fn validate_state(
        &self,
        state_transition: &StateTransition,
        execution_context: &StateTransitionExecutionContext,
    ) -> Result<ConsensusValidationResult<StateTransitionAction>, ProtocolError> {
        self.state_validator
            .validate(state_transition, execution_context)
            .await
    }

    pub async fn apply(&self, state_transition: &StateTransition) -> Result<(), ProtocolError> {
        self.apply_state_transition.apply(state_transition).await
    }
}

pub struct ValidateOptions {
    pub basic: bool,
    pub signature: bool,
    pub fee: bool,
    pub state: bool,
}

impl Default for ValidateOptions {
    fn default() -> Self {
        Self {
            basic: true,
            signature: true,
            fee: true,
            state: true,
        }
    }
}
